default-fields/0000-default-fields.md

```diff
@@ -6,31 +6,206 @@
+    a: &'static str,
+    b: bool,
+    c: i32,
+    d: Vec<i32>
+}
+
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+    ..Default::default()
+}
+```
+
+However, this still requires an already initialised struct after the `..`.
+It also isn't valid if the struct has private fields.
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders.
+The problem with a constructor is that you either need one for each combination of fields a caller can supply
+or expect callers to override values after initialisation, which is inefficient.
+Builders enable more advanced initialisation, but need additional boilerplate.
 
 # Detailed design
 [design]: #detailed-design
 
-This is the bulk of the RFC. Explain the design in enough detail for somebody familiar
-with the language to understand, and for somebody familiar with the compiler to implement.
-This should get into specifics and corner-cases, and include examples of how the feature is used.
+With field defaults a caller can initialise a struct with default values without needing builders
+or a constructor function:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+    d: Vec<i32> = Vec::default()
+}
+
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+}
+```
+
+Field defaults are sugar for the 'real' initialiser, where values for missing fields are added with the supplied default expression.
+
+Supplied field values take precedence over field defaults:
+
+```rust
+// `b` is `false`, even though the field default is `true`
+let foo = Foo {
+    a: "Hello",
+    b: false,
+    c: 42,
+};
+```
+
+Supplied field values in functional updates take precedence over field defaults:
+
+```rust
+// `b` is `false`, even though the field default is `true`
+let foo = Foo {
+    a: "Hello",
+    c: 0,
+    ..Foo {
+        b: false
+    }
+};
+```
+
+The field default should only be used when the caller doesn't supply a value,
+to avoid unnecessarily assigning values.
+
+When deriving `Default`, field defaults are used instead of the type default.
+
+```rust
+#[derive(Default)]
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+}
+
+// `b` is `true`, even though `bool::default()` is `false`
+let foo = Foo::default();
+```
+
+## Allowable Values
+
+Field defaults use the same syntax as `const`s.
+So the type must be supplied, and the value must be a compile-time expression, or a call to `Default::default()`.
+Default field values are expected to be cheap to produce and have basic support for collection types like `Vec`.
 
 # Drawbacks
 [drawbacks]: #drawbacks
 
-Why should we *not* do this?
+Field defaults are limited to `const` expressions and calls to `default()`.
+This means there are values that can't be used as defaults, such as a `Vec` with 3 elements.
+
+Allowing functionality to be injected into data initialisation through abuse of `Default` means struct literal initialisation 
+is no longer guaranteed to be pure.
```

Comment:
I need to think about this more before I can clarify it better. I like that Rust separates data from functionality, it makes things easy to reason about and compose. Allowing arbitrary functionality in the form of `default()` that isn't really about initialisation to sit with the data feels like it messes with that strict separation.

That's all quite fluffy, so I need to focus it better. Do you have any thoughts?